### DISCUSSION TITLE
Scroll Bar Weird Behaviour

### DISCUSSION THREAD

**Discussion by steve09014098** [AvaloniaUI/Avalonia#15435] [Q&A]
I have a weird issue with scrollbars that I cannot solve... Any help would be appreciated!

- Avalonia NuGet package version 11.0.10
- Using Fluent Theme with custom styles
- Visual Studio 2022
- C# / Dot Net 8
- Windows 11

I have styled scroll bars with the following styles which look good, but unfortunately I am seeing some very odd behaviour when the content it is scrolling is large (5000 rows in a data grid for example).

I boiled this down to a really simple thing where I simply declare my scroll bar in XAML without content to replicate it:

https://github.com/AvaloniaUI/Avalonia/assets/28947683/b2c7519a-1159-4fc6-ba4f-498bccae9a66

Essentially it seems like the thumb cannot keep up with the mouse and then somehow overflows?

The scroll bar is declared as:

`              
              <Border Width="1632"
                      Margin="5"
                      BorderBrush="Gray"
                      BorderThickness="1"
                      CornerRadius="5">
                  <Grid Margin="5" RowDefinitions="Auto,Auto,Auto,Auto,Auto,Auto,Auto,Auto,Auto,Auto,Auto,Auto,Auto,Auto">
                          <ScrollBar Height="1000"
                                     HorizontalAlignment="Center"
                                     AllowAutoHide="False"
                                     Maximum="5000"
                                     Minimum="0"
                                     Orientation="Vertical" />
                      </StackPanel>
                  </Grid>
              </Border>
`

The Scroll Styles are:

`
<Styles xmlns="https://github.com/avaloniaui" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">

    
    <Style Selector="ScrollViewer">
        <Setter Property="Background" Value="Transparent" />
        <Setter Property="Foreground" Value="{DynamicResource MainForegroundBrush}" />
        <Setter Property="AllowAutoHide" Value="True" />
        <Setter Property="HorizontalScrollBarVisibility" Value="Disabled" />
        <Setter Property="VerticalScrollBarVisibility" Value="Auto" />

        
        <Setter Property="Template">
            <ControlTemplate>
                <Grid ColumnDefinitions="*,Auto" RowDefinitions="*,Auto">

                    
                    <ScrollBar Name="PART_VerticalScrollBar"
                               Grid.Column="1"
                               Width="20"
                               Background="{TemplateBinding Background}"
                               Foreground="{TemplateBinding Foreground}"
                               Orientation="Vertical"
                               ZIndex="1" />

                    
                    <ScrollContentPresenter Name="PART_ContentPresenter"
                                            Grid.Row="0"
                                            Grid.Column="0"
                                            Padding="{TemplateBinding Padding}"
                                            Background="{TemplateBinding Background}"
                                            HorizontalSnapPointsAlignment="{TemplateBinding HorizontalSnapPointsAlignment}"
                                            HorizontalSnapPointsType="{TemplateBinding HorizontalSnapPointsType}"
                                            ScrollViewer.IsScrollInertiaEnabled="{TemplateBinding IsScrollInertiaEnabled}"
                                            VerticalSnapPointsAlignment="{TemplateBinding VerticalSnapPointsAlignment}"
                                            VerticalSnapPointsType="{TemplateBinding VerticalSnapPointsType}"
                                            ZIndex="0">
                        <ScrollContentPresenter.GestureRecognizers>
                            <ScrollGestureRecognizer CanHorizontallyScroll="{Binding CanHorizontallyScroll, ElementName=PART_ContentPresenter}"
                                                     CanVerticallyScroll="{Binding CanVerticallyScroll, ElementName=PART_ContentPresenter}"
                                                     IsScrollInertiaEnabled="{Binding (ScrollViewer.IsScrollInertiaEnabled), ElementName=PART_ContentPresenter}" />
                        </ScrollContentPresenter.GestureRecognizers>
                    </ScrollContentPresenter>

                    
                    <ScrollBar Name="PART_HorizontalScrollBar"
                               Grid.Row="1"
                               Height="20"
                               Background="{TemplateBinding Background}"
                               Foreground="{TemplateBinding Foreground}"
                               Orientation="Horizontal"
                               ZIndex="1" />

                    
                    <Panel x:Name="PART_ScrollBarsSeparator"
                           Grid.Row="1"
                           Grid.Column="1"
                           Background="{TemplateBinding Background}"
                           Opacity="0"
                           ZIndex="1">
                        <Panel.Transitions>
                            <Transitions>
                                <DoubleTransition Property="Opacity" Duration="0:0:0.1" />
                            </Transitions>
                        </Panel.Transitions>
                    </Panel>
                </Grid>
            </ControlTemplate>
        </Setter>

        
        <Style Selector="^[IsExpanded=true] /template/ Panel#PART_ScrollBarsSeparator">
            <Setter Property="Opacity" Value="1" />
        </Style>

        
        <Style Selector="^[AllowAutoHide=True] /template/ ScrollContentPresenter#PART_ContentPresenter">
            <Setter Property="Grid.ColumnSpan" Value="2" />
            <Setter Property="Grid.RowSpan" Value="2" />
        </Style>
    </Style>

    
    <Style Selector="ScrollBar">

        
        <Style Selector="^:vertical">
            <Setter Property="Template">
                <ControlTemplate>

                    
                    <Grid x:Name="Root" Margin="0,0,0,0">

                        <Border x:Name="VerticalRoot"
                                Background="{TemplateBinding Background}"
                                BorderBrush="{TemplateBinding BorderBrush}"
                                BorderThickness="0">

                            <Grid RowDefinitions="Auto,*,Auto">

                                
                                <Rectangle x:Name="TrackRect"
                                           Grid.RowSpan="3"
                                           Width="{TemplateBinding Width}"
                                           VerticalAlignment="Stretch"
                                           Fill="{DynamicResource MainControlBrush}"
                                           Opacity="0"
                                           StrokeThickness="0">
                                    <Rectangle.Transitions>
                                        <Transitions>
                                            <DoubleTransition Property="Opacity" Duration="0:0:0.1" />
                                        </Transitions>
                                    </Rectangle.Transitions>
                                </Rectangle>

                                
                                <RepeatButton Name="PART_LineUpButton"
                                              Grid.Row="0"
                                              Width="{TemplateBinding Width}"
                                              Height="{TemplateBinding Width}"
                                              HorizontalAlignment="Stretch"
                                              VerticalAlignment="Center"
                                              Background="{TemplateBinding Background}"
                                              Classes="ScrollBarLineButton"
                                              Focusable="False"
                                              Foreground="{TemplateBinding Foreground}"
                                              Opacity="0">
                                    <PathIcon Width="12"
                                              Height="12"
                                              Margin="0"
                                              Background="{TemplateBinding Background}"
                                              Data="{DynamicResource UpChevronIcon}"
                                              Foreground="{DynamicResource MainForegroundBrush}" />
                                </RepeatButton>

                                
                                <Track Grid.Row="1"
                                       Width="{TemplateBinding Width}"
                                       Margin="0"
                                       VerticalAlignment="Stretch"
                                       DeferThumbDrag="{TemplateBinding ScrollViewer.IsDeferredScrollingEnabled}"
                                       IsDirectionReversed="True"
                                       Maximum="{TemplateBinding Maximum}"
                                       Minimum="{TemplateBinding Minimum}"
                                       Orientation="{TemplateBinding Orientation}"
                                       ViewportSize="{TemplateBinding ViewportSize}"
                                       Value="{TemplateBinding Value,
                                                               Mode=TwoWay}">

                                    
                                    <Track.DecreaseButton>
                                        <RepeatButton Name="PART_PageUpButton"
                                                      Margin="0"
                                                      Padding="0"
                                                      Classes="ScrollBarPageButton"
                                                      Focusable="False" />
                                    </Track.DecreaseButton>
                                    <Track.IncreaseButton>
                                        <RepeatButton Name="PART_PageDownButton"
                                                      MinWidth="0"
                                                      Margin="0"
                                                      Padding="0"
                                                      Classes="ScrollBarPageButton"
                                                      Focusable="False" />
                                    </Track.IncreaseButton>

                                    
                                    <Thumb Name="MainThumb"
                                           Width="2"
                                           MinHeight="{DynamicResource ScrollBarSize}"
                                           Margin="0"
                                           Background="{DynamicResource MainAccentBrush}"
                                           Classes="ScrollBarThumb" />
                                </Track>

                                
                                <RepeatButton Name="PART_LineDownButton"
                                              Grid.Row="2"
                                              Width="{TemplateBinding Width}"
                                              Height="{TemplateBinding Width}"
                                              HorizontalAlignment="Stretch"
                                              Background="{TemplateBinding Background}"
                                              Classes="ScrollBarLineButton"
                                              Focusable="False"
                                              Foreground="{TemplateBinding Foreground}"
                                              Opacity="0">
                                    <PathIcon Width="12"
                                              Height="12"
                                              Margin="0"
                                              Background="{TemplateBinding Background}"
                                              Data="{DynamicResource DownChevronIcon}"
                                              Foreground="{DynamicResource MainForegroundBrush}" />
                                </RepeatButton>
                            </Grid>
                        </Border>
                    </Grid>
                </ControlTemplate>
            </Setter>

            
            <Style Selector="^[IsExpanded=true]">

                
                <Style Selector="^ /template/ Rectangle#TrackRect">
                    <Setter Property="Fill" Value="{DynamicResource MainControlBrush}" />
                    <Setter Property="Opacity" Value="1" />
                </Style>

                
                <Style Selector="^ /template/ Thumb#MainThumb">
                    <Setter Property="Width" Value="7" />
                    <Setter Property="Margin" Value="2" />
                </Style>

                
                <Style Selector="^ /template/ RepeatButton#PART_LineUpButton">
                    <Setter Property="Opacity" Value="1" />
                </Style>
                <Style Selector="^ /template/ RepeatButton#PART_LineDownButton">
                    <Setter Property="Opacity" Value="1" />
                </Style>

                <Style Selector="^:disabled">
                    
                    <Style Selector="^ /template/ Rectangle#TrackRect">
                        <Setter Property="Fill" Value="{DynamicResource MainControlBrush}" />
                    </Style>
                    
                    <Style Selector="^ /template/ Thumb#MainThumb">
                        <Setter Property="Background" Value="{DynamicResource MainForegroundBrush}" />
                    </Style>
                </Style>
            </Style>
        </Style>

        
        <Style Selector="^:horizontal">
            <Setter Property="Template">
                <ControlTemplate>

                    
                    <Grid x:Name="Root" Margin="0,0,0,0">

                        <Border x:Name="HorizontalRoot"
                                Background="{TemplateBinding Background}"
                                BorderBrush="{TemplateBinding BorderBrush}"
                                BorderThickness="0">

                            
                            <Grid ColumnDefinitions="Auto,*,Auto">

                                
                                <Rectangle x:Name="TrackRect"
                                           Grid.ColumnSpan="3"
                                           Height="{TemplateBinding Height}"
                                           HorizontalAlignment="Stretch"
                                           Fill="{DynamicResource MainControlBrush}"
                                           Opacity="0"
                                           StrokeThickness="0">
                                    <Rectangle.Transitions>
                                        <Transitions>
                                            <DoubleTransition Property="Opacity" Duration="0:0:0.1" />
                                        </Transitions>
                                    </Rectangle.Transitions>
                                </Rectangle>

                                
                                <RepeatButton Name="PART_LineUpButton"
                                              Grid.Column="0"
                                              Width="{TemplateBinding Height}"
                                              Height="{TemplateBinding Height}"
                                              HorizontalAlignment="Center"
                                              VerticalAlignment="Stretch"
                                              Background="{TemplateBinding Background}"
                                              Classes="ScrollBarLineButton"
                                              Focusable="False"
                                              Foreground="{TemplateBinding Foreground}"
                                              Opacity="0">
                                    <PathIcon Width="12"
                                              Height="12"
                                              Margin="0"
                                              Background="{TemplateBinding Background}"
                                              Data="{DynamicResource LeftChevronIcon}"
                                              Foreground="{DynamicResource MainForegroundBrush}" />
                                </RepeatButton>

                                
                                <Track Grid.Column="1"
                                       Height="{TemplateBinding Height}"
                                       Margin="0"
                                       HorizontalAlignment="Stretch"
                                       DeferThumbDrag="{TemplateBinding ScrollViewer.IsDeferredScrollingEnabled}"
                                       IsDirectionReversed="False"
                                       Maximum="{TemplateBinding Maximum}"
                                       Minimum="{TemplateBinding Minimum}"
                                       Orientation="{TemplateBinding Orientation}"
                                       ViewportSize="{TemplateBinding ViewportSize}"
                                       Value="{TemplateBinding Value,
                                                               Mode=TwoWay}">

                                    
                                    <Track.DecreaseButton>
                                        <RepeatButton Name="PART_PageUpButton"
                                                      Margin="0"
                                                      Padding="0"
                                                      Classes="ScrollBarPageButton"
                                                      Focusable="False" />
                                    </Track.DecreaseButton>
                                    <Track.IncreaseButton>
                                        <RepeatButton Name="PART_PageDownButton"
                                                      Margin="0"
                                                      Padding="0"
                                                      Classes="ScrollBarPageButton"
                                                      Focusable="False" />
                                    </Track.IncreaseButton>

                                    
                                    <Thumb Name="MainThumb"
                                           Height="2"
                                           MinWidth="{DynamicResource ScrollBarSize}"
                                           Margin="0"
                                           Background="{DynamicResource MainAccentBrush}"
                                           Classes="ScrollBarThumb" />
                                </Track>

                                
                                <RepeatButton Name="PART_LineDownButton"
                                              Grid.Column="2"
                                              Width="{TemplateBinding Height}"
                                              Height="{TemplateBinding Height}"
                                              VerticalAlignment="Stretch"
                                              Background="{TemplateBinding Background}"
                                              Classes="ScrollBarLineButton"
                                              Focusable="False"
                                              Foreground="{TemplateBinding Foreground}"
                                              Opacity="0">
                                    <PathIcon Width="12"
                                              Height="12"
                                              Margin="0"
                                              Background="{TemplateBinding Background}"
                                              Data="{DynamicResource RightChevronIcon}"
                                              Foreground="{DynamicResource MainForegroundBrush}" />
                                </RepeatButton>
                            </Grid>
                        </Border>
                    </Grid>
                </ControlTemplate>
            </Setter>

            
            <Style Selector="^[IsExpanded=true]">

                
                <Style Selector="^ /template/ Rectangle#TrackRect">
                    <Setter Property="Fill" Value="{DynamicResource MainControlBrush}" />
                    <Setter Property="Opacity" Value="1" />
                </Style>

                
                <Style Selector="^ /template/ Thumb#MainThumb">
                    <Setter Property="Height" Value="7" />
                    <Setter Property="Margin" Value="2" />
                </Style>

                
                <Style Selector="^ /template/ RepeatButton#PART_LineUpButton">
                    <Setter Property="Opacity" Value="1" />
                </Style>
                <Style Selector="^ /template/ RepeatButton#PART_LineDownButton">
                    <Setter Property="Opacity" Value="1" />
                </Style>
            </Style>

            
            <Style Selector="^:disabled">
                
                <Style Selector="^ /template/ Rectangle#TrackRect">
                    <Setter Property="Fill" Value="{DynamicResource MainControlBrush}" />
                </Style>
                
                <Style Selector="^ /template/ Thumb#MainThumb">
                    <Setter Property="Background" Value="{DynamicResource MainForegroundBrush}" />
                </Style>
            </Style>
        </Style>
    </Style>

    
    <Style Selector="Thumb.ScrollBarThumb">
        <Setter Property="Background" Value="{DynamicResource MainAccentBrush}}" />
        <Setter Property="BorderThickness" Value="0" />
        <Setter Property="CornerRadius" Value="10" />
        <Setter Property="Cursor" Value="Hand" />

        
        <Setter Property="Template">
            <Setter.Value>
                <ControlTemplate>
                    
                    <Border Background="{TemplateBinding Background}" CornerRadius="{TemplateBinding CornerRadius}" />
                </ControlTemplate>
            </Setter.Value>
        </Setter>
    </Style>

    
    <Style Selector="RepeatButton.ScrollBarPageButton">
        <Setter Property="Background" Value="Transparent" />
        <Setter Property="VerticalAlignment" Value="Stretch" />
        <Setter Property="HorizontalAlignment" Value="Stretch" />
        <Setter Property="MinHeight" Value="0" />
        <Setter Property="MinWidth" Value="0" />
        <Setter Property="Opacity" Value="0" />

        
        <Setter Property="Template">
            <ControlTemplate>
                <Border Background="{TemplateBinding Background}" />
            </ControlTemplate>
        </Setter>
    </Style>

    
    <Style Selector="RepeatButton.ScrollBarLineButton">
        <Setter Property="MinHeight" Value="0" />
        <Setter Property="MinWidth" Value="0" />
        <Setter Property="Padding" Value="0" />
        <Setter Property="Margin" Value="0" />
        <Setter Property="BorderThickness" Value="0" />
        <Setter Property="Foreground" Value="{DynamicResource MainForegroundBrush}" />

        
        <Setter Property="Template">
            <ControlTemplate>
                <ContentPresenter x:Name="PART_ContentPresenter"
                                  Content="{TemplateBinding Content}"
                                  Foreground="{TemplateBinding Foreground}" />
            </ControlTemplate>
        </Setter>

        
        <Style Selector="^:pointerover">
            <Style Selector="^ /template/ ContentPresenter">
                <Setter Property="Foreground" Value="{DynamicResource MainAccentBrush}" />
                <Setter Property="Opacity" Value="0.7" />
            </Style>
        </Style>

        
        <Style Selector="^:pressed">
            <Style Selector="^ /template/ ContentPresenter">
                <Setter Property="Foreground" Value="{DynamicResource MainAccentBrush}" />
                <Setter Property="Opacity" Value="1" />
            </Style>
        </Style>

        
        <Style Selector="^:disabled">
            <Style Selector="^ /template/ ContentPresenter">
                <Setter Property="Background" Value="{DynamicResource MainControlBrush}" />
            </Style>
        </Style>
    </Style>
</Styles>
`

**Comment by thevortexcloud** [comment]
If you are using virtualisation then scrolling in general can get fairly weird. If that's what is happening here I am not sure.

**Reply by steve09014098** [reply]
I turned virtualization off but it made no effect to the scrolling :-(

**Reply by yunglook** [reply]
@steve09014098 Did you manage to find a solution for the issue?
